Filesystem-based Teleport storage backend

kontsevoy opened the pull request:
### Changes

There are 2 parts to this PR:

1. Removed unused code and methods from the `backend.Backend` interface, namely `GetValAndTTL()` and `TouchVal()`.
2. Implemented filesystem-based storage back-end with identical behavior to Bolt.

NOTE: the new back end is not plugged in into production code. It is there to be merged later as part of 1.5 milestone.

### The Benefit

From the product perspective, the value proposition of this back-end is:

* Provides HA for people without operational expertise with etcd or Dynamo. They can use network-backed FS (either Gluster/NFS kind or enterprise storage) in combination with multiple auth servers or (even better) with a single auto-migrating auth server.

* Reduced system requirements (Bolt did not work properly on some filesystems).

kontsevoy committed:
Removed some dead code

kontsevoy committed:
Started implemented fs-based backend

kontsevoy committed:
Removed backend.GetValAndTTL()

It was never used anywhere in actual Teleport code.

kontsevoy committed:
Removed backend.TouchVal()

It was never actually used in Teleport

kontsevoy committed:
Finished implementing filesystem-based backend

With full test coverage

kontsevoy committed:
Code cleanup and live testing of fs backend.

kontsevoy committed:
Fixed data race in tests

Note: had to add my own implementation of TimeSource interface because
it will take some time to get Mailgun team to accept my PR into their
timetools package.

klizhentas commented on lib/backend/fs/impl.go:
```diff
@@ -0,0 +1,310 @@
+	log.Debugf("fs.CreateVal(%s/%s) '%v'", strings.Join(bucket, "/"), key, string(val))
+	// do not allow keys that start with a dot
+	if key[0] == '.' {
+		return trace.BadParameter("Invalid key: '%s'. Key names cannot start with '.'", key)
```
error names should start with lowercase

kontsevoy commented:
retest please

kontsevoy commented:
ok to test

klizhentas commented on lib/backend/fs/impl.go:
```diff
@@ -0,0 +1,310 @@
+		bk.DeleteKey(bucket, key)
+		return nil, trace.NotFound("key '%s' is not found", key)
+	}
+	return ioutil.ReadFile(path.Join(dirPath, key))
```
I would suggest using `https://github.com/gravitational/teleport/blob/master/lib/utils/fs.go#L79` as it's a bit more detailed in it's errors

klizhentas commented on lib/backend/fs/impl.go:
```diff
@@ -0,0 +1,310 @@
+	if ttl != backend.Forever {
+		go func() {
+			bk.Clock.Sleep(ttl)
+			bk.ReleaseLock(token)
```
it is possible that process crashes here and the lock will stay forever. To prevent this I suggest to use:

https://github.com/coreos/rkt/blob/master/pkg/lock/file.go

kontsevoy committed:
Comments & minor edits

kontsevoy committed:
Addressed PR comments

- Comments
- Error creation
- Moved from Mailgun's frozen time to clockwork
- Made tests more reliable

kontsevoy commented:
retest this please

kontsevoy commented:
@klizhentas I've addressed all of your PR comments. The locks are going to be removed in the next PR, trying to keep the change sets manageable.

kontsevoy commented:
retest please
